linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,258 @@
+
+`threadpool_schedule_work` shall set the current task state to `TASK_INITIALIZING`.
+
+If task has been already initialized, `threadpool_schedule_work` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity by calling `reallocate_threadpool_array`.
+
+  - If reallcate task array fails, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall obtain task information in next available task array index and return zero on success.
+
+`threadpool_schedule_work` shall set the `task_state` to `TASK_WAITING` and then release the shared SRW lock.
+
+`threadpool_schedule_work` shall unblock the `threadpool` semaphore by calling `sem_post`.
+
+`threadpool_schedule_work` shall decrement the count of pending call that are in progress to be executed.
+
+### threadpool_timer_start
+
+`threadpool_timer_start` shall return zero on success. To be implemented.
+
+### threadpool_timer_restart
+
+`threadpool_timer_restart` shall return zero on success. To be implemented.
+
+### threadpool_timer_cancel
+
+`threadpool_timer_cancel` shall stops the timer and waiting for pending callbacks. To be implemented.
+
+### threadpool_timer_destroy
+
+`threadpool_timer_destroy` stops the timer and frees the resources associate with it. To be implemented. 
+
+### threadpool_work_func
+
+```C
+static int threadpool_work_func(void* param);
+```
+
+`threadpool_work_func` shall obtain a work function.
+
+If `param` is `NULL`, `threadpool_work_func` shall fail and return.
+
+`threadpool_work_func` shall get the real time by calling `clock_gettime`.
+
+`threadpool_work_func` shall  shall wait on the semaphore with a time limit of 2 seconds.
+
+`threadpool_work_func` shall acquire the shared SRW lock by calling `srw_lock_acquire_shared`.
+
+`threadpool_work_func` shall get the current task array size and next waiting task consume index.
+
+If consume index has task state `TASK_WAITING`, `threadpool_work_func` shall set the task state to `TASK_WORKING`.
+
+`threadpool_work_func` shall initialize `task_func` and `task_param` and then set the task state to `TASK_NOT_USED`.
+
+`threadpool_work_func` shall release the shared SRW lock by calling `srw_lock_release_shared`.
+
+If `task_param` is not `NULL`, `threadpool_work_func` shall execute it with parameter `task_param`.
+
+### reallocate_threadpool_array
+
+```C
+static int reallocate_threadpool_array(THREADPOOL* threadpool);
+```
+
+`reallocate_threadpool_array` resize the task array size by doubling it.
+
+`reallocate_threadpool_array` shall acquire the SRW lock in exclusive mode by calling `srw_lock_acquire_exclusive`.
+
+`reallocate_threadpool_array` shall get the current size of task array.
+
+If the array size get overflowed after doubled, `reallocate_threadpool_array` shall fail and return a non-zero value .
+
+Otherwise, `reallocate_threadpool_array` shall double the current task array size and return zero in success.
```

Comment:
> and return zero in success. [](http://example.com/codeflow?start=81&length=28)

I think this "success spec and return 0" would come after moving the memory and making sure the gaps are covered.

Reply:
moved success and return 0 to the end